forem/forem
# Accessibility enhancements to navigation tabs

## What type of PR is this? (check all applicable)


## Description

This PR implements the recommendations on #11853 for navigation tabs, ensuring that:

- each instance appears within a `nav` element
- each `nav` element has a unique aria-label to identify the type of navigation
- each navigation link/button is contained within a list so that the grouping can be understood by screen reader users
- the link/button which is currently selected is indicated by `aria-current="page"`

**The pages with the affected tab lists are:**
- the home feed
- the tag index page
- the search results page
- `/mod` (when viewed on a small screen)
- the notifications page (when viewed on a small screen)
- the analytics dashboard and article 'stats' pages
- the v1 and v2 post editors

I've purposefully avoided updates to the tabs usage in the admin section, as this is currently being refactored as part of the admin v2 navigation.  

**Also covered in this PR (bonus content!):**
- Updates to the Storybook default story
- Adding a a Storybook story showing usage with buttons
- Updating the notes in Storybook for detail on how to implement accessibly

**What isn't covered in this PR:**

#11853 also contains some additional comments/suggestions on places where we use a select for navigation instead of the tabs. I'm going to split this off into its own issue because it's not strictly related to the tabs component (it appears when tabs don't!) and there there may be some design implications in moving from a select to a dropdown. 


## Related Tickets & Documents

Closes #11853 


## QA Instructions, Screenshots, Recordings

I've tested this on desktop and mobile (android) screens. I don't think there is any device-specific concern, but there are a couple of pages that only show the tabs on a mobile screen size (see above list).

In each case we want to check:

- each set of tabs lives inside a navigation element, and that navigation element has an `aria-label` describing it
- the tabs are contained within a list
- the tabs look/behave as before
- whichever tab has the active indicator also has the `aria-current="page"` attribute 
- all other tabs _either_ have an empty `aria-current` value, or no `aria-current` attribute at all
- when the active tab is changed by clicking on a different link, the `aria-current` attributes update accordingly

I would recommend testing this using a screen reader (e.g. VoiceOver on Safari), although the individual HTML structures and attributes could also be inspected in the dev tools.

**Example dev tools content:**
![screenshot of the home feed with dev tools open - the dev tools shows the 'Feed' is selected, and the appropriate aria-current attribute is added](https://user-images.githubusercontent.com/20773163/116665107-3e916480-a991-11eb-8385-4039a43bd2de.png)

**Example screen reader output**
Here's some example output from VoiceOver on Safari for the Home Feed - notice the navigation has a unique name, and the "current page" is announced when scrolling through the tab links:


https://user-images.githubusercontent.com/20773163/116665702-fd4d8480-a991-11eb-9ac1-d6df907a86b0.mp4




### UI accessibility concerns?

This PR introduces some aria properties that can only be tested via screen reader, or short of that, by inspecting the element manually in dev tools (the user experience is best tested by screen reader!).

Although these attributes are normally 'hidden' to most manual testing, the Cypress tested added will make sure these don't regress in future.

## Added tests?

- [X] Yes - I've added Cypress tests for each place this appears. It's a lot of new tests but they all run very quickly, as not much needs to be checked. Despite adding a fair few tests to the suite, I think they are worth having as aria attributes like 'aria-current' are so easily missed in manual tests.

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
      or in a [forem.dev](http://forem.dev) post
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

N/A

## [optional] What gif best describes this PR or how it makes you feel?

![Homer pressing the tab key expecting a Tab soda](https://i.giphy.com/media/aYzxVt2lMrZXW/giphy.webp)
